### PR TITLE
Chore: Replace `worksWithLocalFiles` with `scope`

### PR DESCRIPTION
**Pull request by sarvaje**
Chore: Ignore rules if the scope doesn't apply

## Pull request checklist

Make sure you:

- [x] Signed the [Contributor License Agreement](https://cla.js.foundation/sonarwhal/sonarwhal)
- [x] Followed the [commit message guidelines](https://sonarwhal.com/docs/contributor-guide/contributing/pull-requests/#commitmessages)

For non-trivial changes, please make sure you also:

- [ ] Added/Updated related documentation.
- [ ] Added/Updated related tests.

### CODE REVIEW

**Comment by sarvaje** [comment]
@sonarwhal/core it is in work-in-progress, but if you can take a look that would be great :)

File: packages/rule-no-vulnerable-javascript-libraries/src/rule.ts [hunk 1]
```diff
@@ -222,7 +223,7 @@ const rule: IRuleBuilder = {
             },
             type: 'object'
         }],
-        worksWithLocalFiles: false
+        scope: Scope.site
```
Comment (alrra) [inline packages/rule-no-vulnerable-javascript-libraries/src/rule.ts]: Is there a reason why this cannot be `any` (or at least a TODO)?

File: packages/sonarwhal/src/lib/types/rules.ts [hunk 2]
```diff
@@ -15,7 +16,7 @@ export interface IRuleMetadata {
     /** The schema the rule configuration must follow in order to be valid */
     schema: Array<any>; // TODO: this shouldn't be an Array of any
     /** If the rule works with local resources (file://...) */
```
Comment (alrra) [inline packages/sonarwhal/src/lib/types/rules.ts]: This comment needs to be updated.

File: packages/rule-no-vulnerable-javascript-libraries/src/rule.ts (hunk 1 above)
Comment (sarvaje) [reply packages/rule-no-vulnerable-javascript-libraries/src/rule.ts]: I just change the current value to its equivalent, but you are right, after we add jsdom to the local connector, this should work with local file. Changing it to any!

File: packages/rule-amp-validator/src/rule.ts [hunk 3]
```diff
@@ -8,6 +8,7 @@ import { Category } from 'sonarwhal/dist/src/lib/enums/category';
 import { debug as d } from 'sonarwhal/dist/src/lib/utils/debug';
 import { IRule, IRuleBuilder, ITargetFetchEnd } from 'sonarwhal/dist/src/lib/types';
 import { RuleContext } from 'sonarwhal/dist/src/lib/rule-context';
+import { Scope } from 'sonarwhal/dist/src/lib/enums/scope';
```
Comment (molant) [inline packages/rule-amp-validator/src/rule.ts]: I think `RuleScope` is more meaningful.

File: packages/rule-no-vulnerable-javascript-libraries/src/rule.ts (hunk 1 above)
Comment (molant) [reply packages/rule-no-vulnerable-javascript-libraries/src/rule.ts]: This should be the same as the `axe` rule. You have a comment in there so maybe just have that comment for now. I think the part to inject JS is going to be a bit more complicated in the local connector.

File: packages/sonarwhal/src/lib/cli/rules/common.ts [hunk 4]
```diff
@@ -224,7 +225,7 @@ export class NewRule implements INewRule {
         };
         this.useCase[ruleData.useCase] = true;
         this.prefix = type === QuestionsType.core ? '../../' : 'sonarwhal/dist/src/lib/';
-        this.worksWithLocalFiles = type !== QuestionsType.core;
+        this.scope = (type !== QuestionsType.core) ? Scope.any : Scope.site;
```
Comment (molant) [inline packages/sonarwhal/src/lib/cli/rules/common.ts]: This check is weird, especially once #806 is implemented. This could be a new question that pops depending on the type of rule the user has decided (interop, request, third party service, etc.) or we just ask it directly.

**Comment by molant** [comment]
@sarvaje just a reminder to do a rebase sooner rather than later.

**Comment by molant** [comment]
The [`misc.isHTML()`](https://github.com/sonarwhal/sonarwhal/blob/229b3de5cc555688972bec3dff7de65112ebfa81/packages/sonarwhal/src/lib/utils/misc.ts#L79) function should be updated to not assume blindly that local files are HTML.

**Comment by sarvaje** [comment]
> @sarvaje just a reminder to do a rebase sooner rather than later.

Yes, I have already had that fight...


>The misc.isHTML() function should be updated to not assume blindly that local files are HTML.

Ok, I will take a look.

**Comment by molant** [comment]
Rules checked that are passing (or not)

* [ ] amp-validator
* [ ] apple-touch-icons
* [ ] axe
* [ ] content-type
* [ ] disown-opener
* [ ] highest-available-document-mode
* [ ] html-checker
* [ ] http-cache
* [ ] http-compression
* [ ] image-optimization-cloudinary
* [ ] manifest-app-name
* [ ] manifest-exists
* [ ] manifest-file-extension
* [ ] manifest-is-valid
* [ ] meta-charset-utf-8
* [ ] meta-viewport
* [ ] no-disallowed-headers
* [ ] no-friendly-error-pages
* [ ] no-html-only-headers
* [ ] no-http-redirects
* [ ] no-protocol-relative-urls
* [ ] no-vulnerable-javascript-libraries
* [ ] performance-budget
* [ ] ssllabs
* [ ] strict-transport-security
* [ ] validate-set-cookie-header
* [x] x-content-type-options

**Comment by molant** [comment]
Also I just realized that when running the tests in each package folder the `tests` folder isn't ignored. We have to look into this.

**Comment by molant** [comment]
> Also I just realized that when running the tests in each package folder the `tests` folder isn't ignored. We have to look into this.

This is fixed now via https://github.com/sonarwhal/sonarwhal/commit/e93bb20531885ea768bb5a6ef30bdfc4333f2d28

File: package.json [hunk 5]
```diff
@@ -56,6 +56,6 @@
     "test-on-travis": "yarn lint && yarn build && nyc ava --concurrency=2 --timeout=2m"
   },
   "workspaces": [
-    "packages/!(rule-amp-validator)"
+    "packages/*"
```
Comment (alrra) [inline package.json]: @sarvaje Does `rule-amp-validator` now work?

Comment (sarvaje) [reply package.json]: it is working on windows

Comment (molant) [reply package.json]: Not working for me :/

Comment (molant) [reply package.json]: Even if we remove the `rule-amp-validator` folder we will have problems because rules have a `devDependency` on `sonarwhal: 0.24.0`. The only solution I've found for now in Windows machines is to use bash.
We should revert @sarvaje's change to continue ignoring `rule-amp-validator` so once we publish the first version we can still use the regular prompt.

Comment (sarvaje) [reply package.json]: I don't remember how I did it, but `rule-amp-validator` works perfect in my machine using the windows cmd :s

**Comment by sarvaje** [comment]
@sonarwhal/core This is ready to review :)

File: package.json (hunk 5 above)
Comment (alrra) [reply package.json]: > I don't remember how I did it,

@sarvaje Did you use magic again?

File: packages/sonarwhal/src/lib/utils/content-type.ts [hunk 6]
```diff
@@ -382,11 +382,38 @@ const isTextMediaType = (mediaType: string): boolean => {
     return false;
 };
 
+const getType = (mediaType: string) => {
+    if (mediaType.startsWith('image')) {
+        return 'image';
+    }
+
+    if (mediaType.startsWith('font')) {
```
Comment (alrra) [inline packages/sonarwhal/src/lib/utils/content-type.ts]: There is also `application/vnd.ms-fontobject` for EOT fonts.

File: packages/sonarwhal/src/lib/utils/content-type.ts [hunk 7]
```diff
@@ -382,11 +382,38 @@ const isTextMediaType = (mediaType: string): boolean => {
     return false;
 };
 
+const getType = (mediaType: string) => {
+    if (mediaType.startsWith('image')) {
```
Comment (alrra) [inline packages/sonarwhal/src/lib/utils/content-type.ts]: Do we want something special for the `favicon`?

File: packages/rule-performance-budget/src/connections.ts [hunk 8]
```diff
@@ -60,7 +58,8 @@ const getConnections = (): Array<NetworkConfig> => {
 
     const configsText = configContent
         .replace(/#.*?\n/g, '') // remove comments
-        .split(`${EOL}${EOL}`);
+        .replace(/\r\n/g, '\n') // remove \r just in case the file has a windows EOL
```
Comment (alrra) [inline packages/rule-performance-budget/src/connections.ts]: Why was this changed?

Comment (sarvaje) [reply packages/rule-performance-budget/src/connections.ts]: Because the previous code was assuming that a file has the same EOL than the OS, but yesterday that assumption fail :P

File: packages/sonarwhal/src/lib/utils/content-type.ts (hunk 6 above)
Comment (sarvaje) [reply packages/sonarwhal/src/lib/utils/content-type.ts]: ok, I will add it

File: packages/sonarwhal/src/lib/utils/content-type.ts (hunk 7 above)
Comment (sarvaje) [reply packages/sonarwhal/src/lib/utils/content-type.ts]: AFAIK no.
The getType method is for now the simplest method to get types, but we can improve it when we need more things :)

File: package.json (hunk 5 above)
Comment (sarvaje) [reply package.json]: >@sarvaje Did you use magic again? 😛

Hahahahaha it looks like that xD

File: packages/rule-performance-budget/src/connections.ts (hunk 8 above)
Comment (alrra) [reply packages/rule-performance-budget/src/connections.ts]: Ok.

File: packages/rule-no-disallowed-headers/src/rule.ts [hunk 9]
```diff
@@ -143,11 +144,7 @@ const rule: IRuleBuilder = {
 
         loadRuleConfigs();
 
-        return {
-            'fetch::end': validate,
-            'manifestfetch::end': validate,
-            'targetfetch::end': validate
-        };
+        return { 'fetch::end::*': validate };
```
Comment (alrra) [inline packages/rule-no-disallowed-headers/src/rule.ts]: 💜

File: package.json (hunk 5 above)
Comment (sarvaje) [reply package.json]: So, just to double check before do anything. I have to keep this `"packages/!(rule-amp-validator)"`, right?

**Comment by alrra** [comment]
> So, just to double check before do anything. I have to keep this "packages/!(rule-amp-validator)", right?

Until we make it work, I think we should. @molant ?

---

Also, related (as shared by @molant):  https://github.com/yarnpkg/yarn/issues/4918

File: packages/sonarwhal/tests/lib/cli/parsers/new-parser.ts [hunk 10]
```diff
@@ -70,11 +70,14 @@ test.serial('It should create a new core parser.', async (t) => {
     };
     const parserEventsResult = {
         again: false,
-        event: 'fetch::end'
+        event: 'fetch::end::*'
     };
     const sandbox = sinon.sandbox.create();
 
-    sandbox.stub(misc, 'findPackageRoot').returns(path.join(__dirname, '..', '..', '..', '..', '..'));
+    const packageRoot = path.join(__dirname, '..', '..', '..', '..', '..');
```
Comment (alrra) [inline packages/sonarwhal/tests/lib/cli/parsers/new-parser.ts]: I think this can just be `../../../../../` ?

File: packages/sonarwhal/tests/lib/cli/parsers/new-parser.ts [hunk 11]
```diff
@@ -114,8 +117,10 @@ test.serial('It should create a new core parser with no duplicate events.', asyn
         event: 'element::'
     };
     const sandbox = sinon.sandbox.create();
+    const packageRoot = path.join(__dirname, '..', '..', '..', '..', '..');
```
Comment (alrra) [inline packages/sonarwhal/tests/lib/cli/parsers/new-parser.ts]: Same as at [line 77](https://github.com/sonarwhal/sonarwhal/pull/816/files?#diff-570dd06bb789183b45ec7e43bc8d9958R77).

File: packages/sonarwhal/src/lib/utils/content-type.ts (hunk 6 above)
Comment (alrra) [reply packages/sonarwhal/src/lib/utils/content-type.ts]: @sarvaje Can you add this? Thanks!

**Comment by sarvaje** [comment]
@alrra done!

**Comment by alrra** [comment]
Merged in https://github.com/sonarwhal/sonarwhal/commit/b3196ea9c93a65a23399103c50b461b5e5e4358b. 🎉

**Comment by sarvaje** [comment]
yeah!
